File: task2/characters_replacement_encryption.cpp
#include "characters_replacement_encryption.h"

#include <stdexcept>
#include <cstring>

void CharactersReplacementEncryption::Parse(FILE *f) {
    if (fscanf(f, "%d", &replacement_size) != 1 || replacement_size < 0) {
        throw std::invalid_argument("");
    }
    char originals[MAX_VALID_CHAR_CODE + 1];
    for (char & original : originals) {
        original = ' ';
    }

    for (int i = 0; i < replacement_size; ++i) {
        char original, rpl;
        if (fscanf(f, " %c %c", &original, &rpl) != 2 || !is_valid_symbol(original) ||
            !is_valid_symbol(rpl) || originals[rpl] != ' ') {
            throw std::invalid_argument("");
        }
        originals[rpl] = original;
    }

    size_t len;
    if (fscanf(f, "%lu", &len) != 1 || !len) {
        throw std::invalid_argument("");
    }

    encrypted_string = new char[len + 1];
    try {
        if (fscanf(f, "%s", encrypted_string) != 1 || strlen(encrypted_string) != len) {
            throw std::invalid_argument("");
        }

        size_t code_sum = 0;
        for (size_t i = 0; i < len; ++i) {
            if (!is_valid_symbol(encrypted_string[i]) || originals[encrypted_string[i]] == ' ') {
                throw std::invalid_argument("");
            }
            code_sum += originals[encrypted_string[i]];
        }
        key = double(code_sum) / double(len);
        replacement = new CharReplacement[replacement_size];
        for (int i = 0, j = 0; i <= MAX_VALID_CHAR_CODE; ++i) {
            if (originals[i] != ' ') {
                replacement[j].original = originals[i];
                replacement[j].replacement = (char) i;
                ++j;
            }
        }
    } catch (...) {
        delete[] encrypted_string;
        encrypted_string = nullptr;
        throw;
    }
}

void CharactersReplacementEncryption::Print(FILE *f) const {
    fprintf(f, "Characters Replacement Encryption\nReplacements: ");
    if (replacement_size) {
        fprintf(f, "%c->%c", replacement[0].original, replacement[0].replacement);
        for (int i = 1; i < replacement_size; ++i) {
            CharReplacement *rpl = replacement + i;
            fprintf(f, ", %c->%c", rpl->original, rpl->replacement);
        }
    }
    fprintf(f, "\nEncrypted string: %s", encrypted_string);
}

CharactersReplacementEncryption::~CharactersReplacementEncryption() {
    delete[] replacement;
    delete[] encrypted_string;
}

File: task3/characters_replacement_encryption.py
from encryption import Encryption
from common import *


class CharactersReplacementEncryption(Encryption):
    def __init__(self, arr):
        super().__init__()
        replacement_size = int(arr[1])
        if len(arr) != 4 + replacement_size * 2:
            raise AttributeError
        replacements = arr[2: 2 + 2 * replacement_size]
        if not all(map(is_valid_symbol, replacements)):
            raise AttributeError
        self.replacement = list(zip(replacements[::2], replacements[1::2]))
        self.replacement.sort(key=lambda x: x[1])

        self.encrypted_string = arr[-1]
        if int(arr[-2]) <= 0 or len(self.encrypted_string) != int(arr[-2]):
            raise AttributeError

        decryption = {}
        for x, y in self.replacement:
            if y in decryption.keys():
                raise AttributeError
            decryption[y] = x
        code_sum = 0
        for c in self.encrypted_string:
            if c not in decryption.keys():
                raise AttributeError
            code_sum += ord(decryption[c])
        self.key = code_sum / int(arr[-2])

    def __str__(self):
        return 'Characters Replacement Encryption\nReplacements:\n' + ', '.join(
            map(lambda x: x[0] + '->' + x[1], self.replacement)) + '\nEncrypted string: ' + self.encrypted_string
